src/ucp/wireup/wireup_cm.c

```diff
@@ -229,37 +227,226 @@ static void ucp_cm_client_connect_cb(uct_ep_h ep, void *arg,
+        status = UCS_ERR_UNSUPPORTED;
+        goto err_out;
+    }
+
+    progress_arg = ucs_malloc(sizeof(*progress_arg),
+                              "ucp_cm_client_connect_progress_arg_t");
+    if (progress_arg == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_out;
+    }
+
+    progress_arg->sa_data = ucs_malloc(remote_data->conn_priv_data_length,
+                                        "sa data");
+    if (progress_arg->sa_data == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_free_arg;
+    }
+
+    progress_arg->dev_addr = ucs_malloc(remote_data->dev_addr_length,
+                                        "device address");
+    if (progress_arg->dev_addr == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_free_sa_data;
+    }
+
+    progress_arg->ucp_ep    = ucp_ep;
+    progress_arg->uct_cm_ep = uct_cm_ep;
+    memcpy(progress_arg->dev_addr, remote_data->dev_addr,
+           remote_data->dev_addr_length);
+    memcpy(progress_arg->sa_data, remote_data->conn_priv_data,
+           remote_data->conn_priv_data_length);
+
+    uct_worker_progress_register_safe(worker->uct,
+                                      ucp_cm_client_connect_progress,
+                                      progress_arg, UCS_CALLBACKQ_FLAG_ONESHOT,
+                                      &prog_id);
+    ucp_worker_signal_internal(ucp_ep->worker);
+    return;
+
+err_free_sa_data:
+    ucs_free(progress_arg->sa_data);
+err_free_arg:
+    ucs_free(progress_arg);
+err_out:
+    ucp_ep->flags &= ~UCP_EP_FLAG_LOCAL_CONNECTED;
+    ucp_worker_set_ep_failed(worker, ucp_ep, uct_cm_ep,
+                             ucp_ep_get_cm_lane(ucp_ep), status);
 }
 
-static void ucp_cm_disconnect_cb(uct_ep_h ep, void *arg)
+/*
+ * Internal flush completion callback which is a part of close protocol,
+ * this flush was initiated by remote peer in disconnect callback on CM lane.
+ */
+static void ucp_ep_cm_disconnect_flushed_cb(ucp_request_t *req)
 {
-    ucs_error("UCP close protocol is not completely implemented");
+    ucp_ep_h ucp_ep = req->send.ep;
+
+    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+    ucp_ep_cm_disconnect_cm_lane(ucp_ep);
+    ucp_request_complete_send(req, UCS_OK);
+    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+}
+
+static unsigned ucp_ep_cm_remote_disconnect_progress(void *arg)
+{
+    ucp_ep_h ucp_ep = arg;
+    void *req;
+
+    ucs_trace("ep %p: flags %xu cm_remote_disconnect_progress", ucp_ep,
+              ucp_ep->flags);
+
+    ucs_assert(ucp_ep_get_cm_uct_ep(ucp_ep) != NULL);
+
+    ucs_assert(ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED);
+    if (ucs_test_all_flags(ucp_ep->flags, UCP_EP_FLAG_CLOSED |
+                                          UCP_EP_FLAG_CLOSE_REQ_VALID)) {
+        ucp_request_complete_send(ucp_ep_ext_gen(ucp_ep)->close_req.req, UCS_OK);
+        return 1;
+    }
+
+    if (ucp_ep->flags & UCP_EP_FLAG_CLOSED) {
```

Comment:
can we unite with the 'if' in line 365

Reply:
nope, this checks that ep is closed (by API) but close req is not valid yet, will be set from scheduled `ucp_ep_close_flushed_callback`